platform/pulpcore/app/models/storage.py

```diff
@@ -101,300 +97,86 @@ def __call__(self, model, name):
-                pass
-
-        return path
-
-    @staticmethod
-    def get_artifact_path(sha256digest):
-        """
-        Determine the absolute path where a file backing the Artifact should be stored.
-
-        Args:
-            sha256digest (str): sha256 digest of the file for the Artifact
-
-        Returns:
-            A string representing the absolute path where a file backing the Artifact should be
-            stored
-        """
-        return os.path.join(settings.MEDIA_ROOT, 'artifact', sha256digest[0:2], sha256digest[2:])
+        if name is None:
+            name = content.name
 
-    @staticmethod
-    def published_metadata_path(model, name):
-        """
-        Get the storage path for published metadata.
-
-        Args:
-            model (pulpcore.app.models.PublishedMetadata): A model instance.
-            name (str): The file name.
-
-        Returns:
-            str: The absolute storage path.
-        """
-        return os.path.join(
-            settings.MEDIA_ROOT,
-            'published',
-            'metadata',
-            str(model.pk),
-            name)
-
-    def get_available_name(self, name, max_length=None):
-        """
-        Get the available absolute path based on the name requested.
-        Required by the Storage API.
-
-        Args:
-            name (str): File name.
-            max_length (int): Maximum length, in characters, of the returned path.
-
-        Returns:
-            str: Available name.
-
-        Raises:
-            ValueError if ``max_length`` exceeded.
-        """
-        if max_length:
-            if len(name) > max_length:
-                raise ValueError('max_length exceeded')
-        return name
-
-    def delete(self, path):
-        """
-        Delete the file at the specified path.
-        Required by the Storage API.
+        if not hasattr(content, 'chunks'):
+            content = File(content, name)
 
-        Args:
-            path (str): An absolute file path.
-        """
-        FileSystem.unlink(path)
-
-    def exists(self, path):
-        """
-        Get whether the file actually exists at the specified path.
-        Required by the Storage API.
-
-        Args:
-            path (str): An absolute file path.
-
-        Returns:
-            bool: True if exists.
-        """
-        return os.path.exists(path)
-
-    def listdir(self, path):
-        """
-        List the content of the directory at the specified path.
-        Required by the Storage API.
-
-        Args:
-            path (str): An absolute directory path.
-
-        Returns:
-            A tuple of two lists: (directories, files).
-        """
-        files = []
-        directories = []
-        for entry in os.listdir(path):
-            if os.path.isdir(os.path.join(path, entry)):
-                directories.append(entry)
+        try:
+            name = self.get_available_name(name, max_length=max_length)
+            return self._save(name, content)
```

Comment:
Are you catching exceptions from this line? If not, it can move to an `else` block.

Reply:
i am catching exceptions in self._save(). self.get_available_name() can get called again during self._save() if the file already exists.